File: days/day-04/solutions/day04.py
import re
import sys

INPUT_FILE = "input.txt"

BIRTH_YEAR = 'byr'
ISSUE_YEAR = 'iyr'
EXPIRATION_YEAR = 'eyr'
HEIGHT = 'hgt'
HAIR_COLOR = 'hcl'
EYE_COLOR = 'ecl'
PASSPORT_ID = 'pid'
COUNTRY_ID = 'cid'

REQUIRED_PROPS = {
    BIRTH_YEAR,
    ISSUE_YEAR,
    EXPIRATION_YEAR,
    HEIGHT,
    HAIR_COLOR,
    EYE_COLOR,
    PASSPORT_ID,
}


class Passport:
    def __init__(self, text: str):
        self.props = {}
        for prop in text.split(" "):
            key, value = prop.split(":")
            self.props[key] = value

    def valid(self, apply_rules: bool):
        if not REQUIRED_PROPS.issubset(self.props.keys()):
            return False

        # Only for part 2
        if not apply_rules:
            return True

        for v in [
            lambda: self.eval_height(),
            lambda: self.eval_eye_color(),
            lambda: self.eval_in_range(BIRTH_YEAR, 1920, 2002),
            lambda: self.eval_in_range(ISSUE_YEAR, 2010, 2020),
            lambda: self.eval_in_range(EXPIRATION_YEAR, 2020, 2030),
            lambda: self.eval_with_regex(HAIR_COLOR, r'^\#[0-9a-f]{6}$'),
            lambda: self.eval_with_regex(PASSPORT_ID, r'^\d{9}$'),
        ]:
            if not v():
                return False

        return True

    def eval_in_range(self, prop_name: str, lower: int, upper: int):
        val = self.props.get(prop_name)
        return lower <= int(val) <= upper

    def eval_with_regex(self, prop_name: int, regex):
        val = self.props.get(prop_name)
        return re.match(regex, val)

    def eval_height(self):
        val = self.props.get(HEIGHT)

        if "in" in val:
            val = val.replace("in", "")
            return 59 <= int(val) <= 76
        if "cm" in val:
            val = val.replace("cm", "")
            return 150 <= int(val) <= 193

        return False

    def eval_eye_color(self):
        val = self.props.get(EYE_COLOR)
        return val in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']


def get_passports():
    data = sys.stdin.read().strip()
    data = data.split("\n\n")
    data = [x.replace("\n", " ") for x in data]

    return [Passport(x) for x in data]


if __name__ == "__main__":
    passports = get_passports()

    # Part 1
    passports = list(filter(lambda x: x.valid(apply_rules=False), passports))
    print(f"Part 1 => {len(passports)}")

    # Part 2
    passports = list(filter(lambda x: x.valid(apply_rules=True), passports))
    print(f"Part 2 => {len(passports)}")

File: days/day-04/solutions/day04.ts
const lines = await Deno.readTextFile('/dev/stdin')

const passportStrings = [...lines.matchAll(/\S+([ \n][\S]+)*/g)]
  .map(([match]) => match);

const passportTokens = passportStrings
  .map(pass => pass.replaceAll(/\n/g, " ").split(" "))

const passportKeyValues = passportTokens
  .map(tokens => tokens.map((token: string) => token.split(":")))

const REQUIRED_VALIDATORS = {
  //  byr (Birth Year) - four digits; at least 1920 and at most 2002.
  byr: (value: string) => !!value.match(/[\d]{4}/)
    && 1920 <= parseInt(value) && parseInt(value) <= 2002
  ,

  //  iyr (Issue Year) - four digits; at least 2010 and at most 2020.
  iyr: (value: string) => !!value.match(/[\d]{4}/)
    && 2010 <= parseInt(value) && parseInt(value) <= 2020
  ,

  //  eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
  eyr: (value: string) => !!value.match(/[\d]{4}/)
    && 2020 <= parseInt(value) && parseInt(value) <= 2030
  ,

  //  hgt (Height) - a number followed by either cm or in:
  //  If cm, the number must be at least 150 and at most 193.
  //  If in, the number must be at least 59 and at most 76.
  hgt: (value: string) => {
    const [,num, unit] = value.match(/(\d\d\d?)(in|cm)/) ?? [];
    if (!num) return false;

    const number = parseInt(num);

    return (unit === "cm" && 150 <= number && number <= 193)
      || (unit === "in" && 59 <= number && number <= 76);
  },

  //  hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
  hcl: (value: string) => !!value.match(/#[a-z0-9]{6}/),

  //  ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
  ecl: (value: string) => !!value.match(/amb|blu|brn|gry|grn|hzl|oth/),

  // pid (Passport ID) - a nine-digit number, including leading zeroes.
  pid: (value: string) => value.length === 9 && !!value.match(/[0]*[1-9][0-9]*/),
} as {
  [key: string]: (value: string) => boolean
}

const REQUIRED_KEYS = Object.keys(REQUIRED_VALIDATORS);

const passportRequiredKeyValues = passportKeyValues
  .map(keyValues => keyValues
    .filter(([key]) => REQUIRED_KEYS.includes(key)))

const [part1, part2] = passportRequiredKeyValues
  .reduce(([part1, part2], requiredKeyValues) => {

    const keys = requiredKeyValues
      .map(([key]) => key)

    const keySet = [...new Set(keys)];

    const passportKeysAreValid = REQUIRED_KEYS
      .every(key => keySet.includes(key));

    const passportValuesAreValid = passportKeysAreValid &&
      requiredKeyValues
        .every(([key, value]) => REQUIRED_VALIDATORS[key](value));

    return [
      part1 + (passportKeysAreValid ? 1:0),
      part2 + (passportValuesAreValid ? 1:0)
    ];
  }, [0,0])

console.log(`${part1}`);
console.log(`${part2}`);
